File: framework/test2/void/186/revword.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
class Solution:
    def reverseWords(self, s: List[str]) -> None:
        if len(s) == 0:
            return
        self.reverse(s, 0, len(s)-1)
        N = len(s)
        i, j = 0, 0
        while j < N:
            if s[j] == ' ':
                self.reverse(s, i, j-1)
                i = j + 1
            j += 1
        if i < j:
            self.reverse(s, i, j-1)

    def reverse(self, s, i, j):
        while i < j:
            temp = s[i]
            s[i] = s[j]
            s[j] = temp
            i += 1
            j -= 1

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().reverseWords)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    # work.setArgumentParser(index, (a) => b)
    # work.resultSerializer = (r) => s
    # work.resultParser = (s) => r
    work.run()

File: framework/test2/void/186/Revword.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Function;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.*;

class Solution {
    public void reverseWords(char[] s) {
        if (s.length == 0) {
            return;
        }

        reverse(s, 0, s.length - 1);

        int N = s.length;
        int i = 0, j = 0;
        while (j < N) {
            if (s[j] == ' ') {
                reverse(s, i, j-1);
                i = j + 1;
            }
            ++j;
        }
        if (i < j) {
            reverse(s, i, j-1);
        }
    }

    void reverse(char[] s, int i, int j) {
        while (i < j) {
            char temp = s[i];
            s[i] = s[j];
            s[j] = temp;
            ++i;
            --j;
        }
    }
}

public class Revword implements Function<String, String> {

    @Override
    public String apply(String text) {
        var work = GenericTestWork.create1u(new Solution()::reverseWords);
        // var work = new TestWork(new Solution(), "reverseWords");
        // work.setValidator((e, r) -> {...});
        work.setCompareSerial(true);
        // work.setArgumentParser(index, a -> { ... });
        // work.setResultParser(r -> { ... });
        // work.setResultSerializer(r -> {...});
        return work.run(text);
    }

    public static void main(String[] args) throws Exception {
        System.out.println(new Revword().apply(Utils.fromStdin()));
    }

}
